hail/python/hail/linalg/utils/misc.py

```diff
@@ -165,51 +166,43 @@ def locus_windows(locus_expr, radius, coord_expr=None):
     if radius < 0:
         raise ValueError(f"locus_windows: 'radius' must be non-negative, found {radius}")
     check_row_indexed('locus_windows', locus_expr)
+
     if coord_expr is None:
-        global_pos_list = locus_expr.global_position().collect()
-        n_loci = len(global_pos_list)
-        global_pos = np.zeros(n_loci, dtype=np.int64)
-        for i, p in enumerate(global_pos_list):
-            if p is None:
-                raise ValueError(f"locus_windows: missing value for 'locus_expr' global position at row {i}")
-            global_pos[i] = p
-        coord = global_pos
-        del global_pos_list
+        coord_expr = locus_expr.position
     else:
         check_row_indexed('locus_windows', coord_expr)
-        global_pos_and_coord =\
-            hl.tuple([locus_expr.global_position(), coord_expr]).collect()  # raises exception if sources differ
-        n_loci = len(global_pos_and_coord)
-
-        global_pos = np.zeros(n_loci, dtype=np.int64)
-        coord = np.zeros(n_loci, dtype=np.float64)
-        for i, x in enumerate(global_pos_and_coord):
-            if x[0] is None:
-                raise ValueError(f"locus_windows: missing value for 'locus_expr' global position at row {i}")
-            global_pos[i] = x[0]
-            if x[1] is None:
-                raise ValueError(f"locus_windows: missing value for 'coord_expr' at row {i}")
-            coord[i] = x[1]
-        del global_pos_and_coord
-
-    if n_loci == 0:
-        return np.zeros(shape=0, dtype=np.int64), np.zeros(shape=0, dtype=np.int64)
 
-    contig_name = locus_expr.dtype.reference_genome.contigs
-    contig_len = locus_expr.dtype.reference_genome.lengths
-    contig_cum_len = np.cumsum([contig_len[name] for name in contig_name])
+    rg = locus_expr.dtype.reference_genome
+    contig_group_expr = hl.agg.group_by(hl.locus(locus_expr.contig, 1, reference_genome=rg), hl.agg.collect(coord_expr))
+
+    # check loci are in sorted order
+    last_pos = hl.fold(lambda a, elt: (hl.case()
+                                         .when(a <= elt, elt)
+                                         .or_error("locus_windows: 'locus_expr' global position must be in ascending order.")),
+                       -1,
+                       hl.agg.collect(hl.case()
+                                        .when(hl.is_defined(locus_expr), locus_expr.global_position())
+                                        .or_error("locus_windows: missing value for 'locus_expr'.")))
+    checked_contig_groups = (hl.case()
+                               .when(last_pos >= 0, contig_group_expr)
+                               .or_error("locus_windows: 'locus_expr' has length 0"))
+
+    src = locus_expr._indices.source
```

Comment:
this bit and below can be:
`checked_contig_groups._aggregation_method(checked_contig_groups, _localize=False)`